apps/cyberstorm-remix/cyberstorm/utils/dapperSingleton.ts

```diff
@@ -0,0 +1,100 @@
+import { DapperTs } from "@thunderstore/dapper-ts";
+import { type RequestConfig } from "@thunderstore/thunderstore-api";
+import { getSessionTools } from "../security/publicEnvVariables";
+import { getCachedRequestPromise } from "./requestCache";
+
+type ConfigFactory = () => RequestConfig;
+
+let currentConfigFactory: ConfigFactory | undefined;
+const requestScopedProxies = new WeakMap<Request, DapperTs>();
+
+function resolveConfigFactory(): ConfigFactory {
+  if (currentConfigFactory) {
+    return currentConfigFactory;
+  }
+
+  const tools = getSessionTools();
+  currentConfigFactory = () => tools.getConfig();
+  return currentConfigFactory;
+}
+
+function updateDapperConfig(factory: ConfigFactory) {
+  currentConfigFactory = factory;
+  if (typeof window !== "undefined" && window.Dapper) {
+    window.Dapper.config = () => factory();
+  }
+}
+
+export function initializeClientDapper(factory?: ConfigFactory) {
+  if (typeof window === "undefined") {
+    return;
+  }
+
+  if (factory) {
+    updateDapperConfig(factory);
+  } else if (!currentConfigFactory) {
+    resolveConfigFactory();
+  }
+
+  if (!window.Dapper) {
+    const resolvedFactory = resolveConfigFactory();
+    window.Dapper = new DapperTs(() => resolvedFactory());
+  }
+}
+
+export function getClientDapper(): DapperTs {
+  if (typeof window === "undefined") {
+    throw new Error("getClientDapper can only run in the browser");
+  }
+
+  if (!window.Dapper) {
+    initializeClientDapper();
+  }
+
+  return window.Dapper;
+}
+
+export function getRequestScopedDapper(request?: Request): DapperTs {
```

Comment:
I'm not sure about the naming here. "RequestScoped" raises more questions than provides answers. `getDapperForRequest` or `getClientLoaderDapper` would be clearer, although they kind of hide the deduplication aspect from the reader of the code. But is that a bad thing? Thoughts?

Also a docstring shortly explaining what this does and when it should be used (and why) would be a good addition.